Core/src/Navigation/NavigationStreamHelper.cpp

```diff
@@ -0,0 +1,130 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Navigation/NavigationStreamHelper.hpp"
+
+#include "Acts/Surfaces/Surface.hpp"
+
+bool Acts::NavigationStreamHelper::initializeStream(
+    NavigationStream& stream, const GeometryContext& gctx,
+    const NavigationStream::QueryPoint& queryPoint,
+    BoundaryTolerance cTolerance) {
+  // Position and direction from the query point
+  const Vector3& position = queryPoint.position;
+  const Vector3& direction = queryPoint.direction;
+
+  // De-duplicate first (necessary to deal correctly with multiple
+  // intersections) - sort them by surface pointer
+  std::sort(stream.candidates.begin(), stream.candidates.end(),
+            [](const NavigationStream::Candidate& a,
+               const NavigationStream::Candidate& b) {
+              return (&a.surface()) < (&b.surface());
+            });
+  // Remove duplicates on basis of the surface pointer
+  stream.candidates.erase(
+      std::unique(stream.candidates.begin(), stream.candidates.end(),
+                  [](const NavigationStream::Candidate& a,
+                     const NavigationStream::Candidate& b) {
+                    return (&a.surface()) == (&b.surface());
+                  }),
+      stream.candidates.end());
+
+  // A container collecting additional valid ones from multi intersections
+  std::vector<NavigationStream::Candidate> miCandidates = {};
+  for (auto& [sIntersection, portal, bTolerance, abortTarget] :
+       stream.candidates) {
+    // Get the surface from the object intersection
+    const Surface* surface = sIntersection.object();
+    // Intersect the surface
+    auto multiIntersection = surface->intersect(
+        gctx, position, direction, cTolerance, s_onSurfaceTolerance);
+
+    // Split them into valid intersections
+    bool miCandidate = false;
+    for (auto& rsIntersection : multiIntersection.split()) {
+      // Skip negative solutions
+      if (rsIntersection.pathLength() < 0.) {
```

Comment:
I usually probe against the `onSurfaceTolerance` instead of `0` because this is the finite space we would still consider to be on surface. Anything beyond that is what we want to encounter

Reply:
ok, sounds good.

Reply:
But that should be negative onSurface tolerance, right?